common/src/main/java/com/microsoft/identity/common/internal/telemetry/Telemetry.java

```diff
@@ -0,0 +1,283 @@
+     *
+     * @param event BaseEvent object
+     * @return the event reference for future properties modification.
+     */
+    public static BaseEvent emit(final BaseEvent event) {
+        if (getInstance().mIsTelemetryEnabled) {
+            //only enqueue the telemetry properties when the telemetry is enabled.
+            getInstance().getRequestMap().add(event.getProperties());
+        }
+
+        return event;
+    }
+
+    public void flush() {
+        if (getInstance().mIsTelemetryEnabled) {
+            flush(DiagnosticContext.getRequestContext().get(CORRELATION_ID));
+        }
+    }
+
+    /**
+     * Flush the telemetry data based on the requestId to the observer.
+     *
+     * @param correlationId The request id should either passed by the client app through the API call
+     *                      or generated by the API dispatcher.
+     */
+    public void flush(@NonNull final String correlationId) {
+        if (!mIsTelemetryEnabled) {
+            Logger.warn(TAG, "Telemetry is disabled.");
+            return;
+        }
+        if (null == mObservers) {
+            Logger.warn(TAG, "No telemetry observer set.");
+            return;
+        }
+
+        if (StringUtil.isEmpty(correlationId)) {
+            Logger.warn(TAG, "No correlation id set.");
+            return;
+        }
+
+        synchronized (this) {
+            //check the configuration
+            if (!mDefaultConfiguration.isDebugEnabled() && mIsDebugging) {
+                return;
+            }
+
+            List<Map<String, String>> finalRawMap = new ArrayList<>();
+
+            for (Map<String, String> event : mTelemetryRawDataMap) {
+                if (correlationId.equalsIgnoreCase(event.get(TELEMETRY_KEY_CORRELATION_ID))) {
+                    finalRawMap.add(applyPiiOiiRule(event));
+                    finalRawMap.remove(event);
+                }
+            }
+
+            //Add the telemetry context to the telemetry data
+            finalRawMap.add(applyPiiOiiRule(mTelemetryContext.getProperties()));
+
+            for (ITelemetryObserver observer : mObservers) {
+                observer.send(finalRawMap);
+            }
+        }
+    }
+
+    private Map<String, String> applyPiiOiiRule(final Map<String, String> properties) {
+        if (mDefaultConfiguration.isPiiEnabled()) {
+            Logger.warn(TAG, "Telemetry PII/OII is enabled by the developer.");
+            return properties;
+        }
+
+        final Map<String, String> nonPiiProperties = new HashMap<>();
+        for (Map.Entry<String, String> entry : properties.entrySet()) {
+            if (!TelemetryPiiOiiRules.getInstance().isPiiOrOii(entry.getKey())) {
+                nonPiiProperties.put(entry.getKey(), entry.getValue());
+            }
+        }
+
+        return nonPiiProperties;
+    }
+
+    /**
+     * API for creating {@link Telemetry} instances.
+     */
+    public static class Builder {
```

Comment:
Because the Telemetry instance needs to be initialized with Context and Configuration as parameters. And I prefer it use Builder instead of an initialize(context, telemetryConfiguration) and having something like
```java
Telemetry telemetry = Telemetry.getInstance();
telemetry.initialize(context, configuration);
```